docs/design/storage-node-automatic-updater.md

```diff
@@ -0,0 +1,143 @@
+# Storage Node Automatic Updater
+
+## Abstract
+
+Automatic Updater is a process that downloads the latest Storage Node binary and replaces the currently running one.
+
+## Background
+
+As more Storage Node Operators join the network we not keep their nodes up to date.
+If a Storage Node doesn't meet the minimum version required by the satellites they will no longer be able to offer services to the network.
+Currently we are using Docker for updates, but due to it's limitations with certain OS-s we need a better solution.
+
+The Updater has several responsibilities:
+
+1. Figure out whether something needs to be updated with gradual rollout/
+1. Safely download the binaries.
+1. Safely update the binaries.
+1. Safely restart the binaries.
+
+## Design
+
+The Updater has several steps it takes, contact version server, download, update, restart.
+
+### Checking for updates
+
+Update check will regularly, with jitter, contact Version Server, which responds with a message:
+
+```json
+{
+  "processes": {
+    "storagenode": {
+      "allowed_version_minimum": "0.3.4",
+      "suggested_version": "0.5.1",
+      "rollout": {
+        "active": true,
+        "target_version": "0.5.2",
+        "rollout_seed": "04123bacde",
+        "rollout_cursor": "40"
+      }
+    }
+  }
+}
+```
+
+When there is a newer version is available it needs to calculate whether it needs to update. To check whether rollout has reached this node it needs to calculate `hash(rollout_seed, node_id) < rollout_cursor`. This exact behavior may differ for canary nodes, which always get the latest version.
+
+* The update check must verify that it is a trusted server.
+* The update check should have a jitter to avoid a stampeding herd. See http://highscalability.com/blog/2012/4/17/youtube-strategy-adding-jitter-isnt-a-bug.html for more information.
```

Comment:
Often, after update storage nodes to DB migration (which may bring the offline for long minutes) or initiate some communication with satellites. If too many nodes update at the same time, this may bring a large part of the network offline, or the satellites may be overloaded with too many requests. We want to avoid this with adding a jitter.